### PR TITLE
Compute and backpropagate error loss and complexity loss separately

### PR DESCRIPTION
**Pull request by mrcslws**
Before this change, complexity loss functions were memory-constrained, since storing the forward pass computation graph uses nearly all available memory. This competition for memory is totally unnecessary; complexity loss can be computed and backpropagated after this computation graph has been freed.

### CODE REVIEW

**Comment by mrcslws** [comment]
I ran and profiled an experiment that uses complexity loss. Everything looks good.

**Review by lscheinkman — APPROVED** [review]
